lib2/client/okta.go

```diff
@@ -0,0 +1,667 @@
+	if err != nil {
+		// log the error to debug and continue
+		log.Debug("func:NewOktaClient ", err)
+	}
+	return &oktaClient, nil
+}
+
+// Gets the Okta session cookie and stores it in the cookie jar used by the
+// http client which is used as the primary authentication mechanism.
+//
+// If a keyring isn't provided to the client then an error will be returned.
+// This error indicates that the session wasn't retrieved and should be handled
+// appropriately.
+func (o *OktaClient) retrieveSessionCookie() (err error) {
+
+	if o.Keyring == nil {
+		return fmt.Errorf("Session NOT retrieved. Reason: Session Backend not defined")
+	}
+	cookieItem, err := (*o.Keyring).Get(o.getSessionCookieKeyringKey())
+	if err == nil {
+		o.CookieJar.SetCookies(o.BaseURL, []*http.Cookie{
+			{
+				Name:  "sid",
+				Value: string(cookieItem.Data),
+			},
+		})
+		log.Debug("Using Okta session: ", string(cookieItem.Data))
+	}
+
+	return
+}
+
+// returns the sesion key that is username and domain aware.
+func (o *OktaClient) getSessionCookieKeyringKey() string {
+	return "okta-session-cookie-" + o.creds.Username + "-" + o.creds.Domain
+}
+
+// Takes a session cookie in the cookie jar and saves it in the keychain,
+// this allows it to be used across invocations where the client making the
+// request is destroyed/created between requests.
+func (o *OktaClient) SaveSessionCookie() (err error) {
+
+	if o.Keyring == nil {
+		return fmt.Errorf("Session NOT saved. Reason: Session Backend not defined")
+	}
+	cookies := o.CookieJar.Cookies(o.BaseURL)
+	for _, cookie := range cookies {
+		if cookie.Name == "sid" {
+			newCookieItem := keyring.Item{
+				Key:                         o.getSessionCookieKeyringKey(),
+				Data:                        []byte(cookie.Value),
+				Label:                       "okta session cookie for " + o.creds.Username,
+				KeychainNotTrustApplication: false,
+			}
+			err = (*o.Keyring).Set(newCookieItem)
+			if err != nil {
+				return
+			}
+			log.Debug("Saving Okta Session Cookie: ", cookie.Value)
+		}
+	}
+	return
+}
+
+// Sends a request to the Okta Sessions API to validate if the session cookie
+// is valid or not. This doesn't always mean that the session can be used for
+// all Okta applications but it does accurately fetch the state of the session.
+func (o *OktaClient) ValidateSession() (sessionValid bool, err error) {
+	var mySessionResponse *http.Response
+	sessionValid = false
+
+	log.Debug("Checking if we have a valid Okta session")
+	mySessionResponse, err = o.Request("GET", "api/v1/sessions/me", url.Values{}, []byte{}, "json", false)
+	if err != nil {
+		return
+	}
+	defer mySessionResponse.Body.Close()
+
+	// https://developer.okta.com/docs/reference/api/sessions/#get-current-session
+	// "if the session is invalid, a 404 Not Found response will be returned."
+	// checking for ok status (200) is adequate to see if the session is still valid.
+	sessionValid = mySessionResponse.StatusCode == http.StatusOK
```

Comment:
I wonder if this should just return an error if the session is invalid, instead of also having this bool

Reply:
I wanted to have a separate `bool` so that we can determine if the request is failing to complete vs the session not being valid. Not sure how important that is but that is why I split it out.

Reply:
I'd say typed errors are a better way to deal with that (along with `errors.Is/As`), but I can see that having a `bool` might be simpler.